BW6-761 curve operations (#3026)
ethereum/EIPs: EIPS/eip-3026.md

##### Encoding of uncompressed points:

Points in both G1 and G2 can be expressed as `(x, y)` affine coordinates, where `x` and `y` are elements of the base field.
Therefore, points in both G1 and G2 are encoded as the byte concatenation of the field element encodings of the `x` and `y` affine coordinates. The total encoding length for a G1/G2 point is thus `192` bytes.

##### Point at infinity encoding:

Also referred as the "zero point". For BW6-761 (`y^2=x^3-1`) and its M-twisted curves (`y^3=x^3+4`), the point with coordinates `(0, 0)` (formal zeros in Fp) is *not* on the curve, and so the encoding of `(0, 0)` is used as a convention to encode the point at infinity.

##### Encoding of scalars for multiplication and multiexponentiation operations:

For multiplication and multiexponentiation operations, a scalar is encoded as `64` bytes by performing BigEndian encoding of the corresponding (unsigned) integer.

Note that the main subgroup order for BW6-761 is actually only `377` bits (`48` bytes), but an encoding of `64` bytes has been chosen to have a `32`-byte-aligned ABI (representable as e.g. `bytes32[2]` or `uint256[2]`).

> The corresponding integer is **not** required to be less than or equal to the main subgroup order.

Reviewer: ```suggestion
The corresponding integer **MAY** be greater than the main subgroup order.
```
RFC-2119

Error cases:
- Any of G2 points being not on the curve must result in error
- Field elements encoding rules apply (obviously)
- Input has invalid length

##### ABI for pairing

Pairing call expects `384*k` bytes as an input, that is interpreted as the byte concatenation of `k` slices. Each slice has the following structure:
- `192` bytes G1 point encoding
- `192` bytes G2 point encoding

Output is `32` bytes representing a boolean:

- `0x0000000000000000000000000000000000000000000000000000000000000001` if the pairing result is equal the to multiplicative identity in the pairing target field; and
- `0x0000000000000000000000000000000000000000000000000000000000000000` otherwise.

Error cases:
- Any of the G1 or G2 points being not on the curve
- Any of the G1 or G2 points being not in the correct subgroup
- Input has invalid length
- Field elements encoding rules apply (obviously)

#### Prevention of DDoS on error handling

> This precompile performs extensive computations and in case of any errors during execution it MUST consume all gas from the the gas schedule for the corresponding operation.

Reviewer: ```suggestion
This precompile performs extensive computations and in case of any errors during execution it **MUST** consume all gas from the the gas schedule for the corresponding operation.
```

```

`max_discount = 174`

##### Pairing operation

Average timing of 1000 random samples (random points in G1 and G2) for different number of pairs with linear lifting:
Base cost of the pairing operation is `XXXXX*k + XXXXX` where `k` is a number of pairs.

> Each point (either G1 or G2) for which subgroup check is requested and performed adds the corresponding G1/G2 multiplication cost to it.

Reviewer: Most of this content should be in the Rationale section.  The specification section should only have the final list of gas costs and/or the algorithm for gas cost calculation (assuming the algorithm is part of consensus like for some of the other cryptographic precompiles).  How you decided on the gas costs is great content for the Rationale section.

There are no backward compatibility questions.

## Test Cases

Due to the large test parameters space we first provide properties that various operations must satisfy. We use additive notation for point operations, capital letters (`P`, `Q`) for points, small letters (`a`, `b`) for scalars. Generator for G1 is labeled as `G`, generator for G2 is labeled as `H`, otherwise we assume random point on a curve in a correct subgroup. `0` means either scalar zero or point of infinity. `1` means either scalar one or multiplicative identity. `group_order` is a main subgroup order. `e(P, Q)` means pairing operation where `P` is in G1, `Q` is in G2.

Requeired properties for basic ops (add/multiply):

- Commutativity: `P + Q = Q + P`
- Additive negation: `P + (-P) = 0`
- Doubling `P + P = 2*P`
- Subgroup check: `group_order * P = 0`
- Trivial multiplication check: `1 * P = P`
- Multiplication by zero: `0 * P = 0`
- Multiplication by the unnormalized scalar `(scalar + group_order) * P = scalar * P`

Required properties for pairing operation:
- Degeneracy `e(P, 0*Q) = e(0*P, Q) = 1`
- Bilinearity `e(a*P, b*Q) = e(a*b*P, Q) = e(P, a*b*Q)` (internal test, not visible through ABI)

> Test vector for all operations are expanded in this [gist](https://gist.github.com/shamatar/506ab3193a7932fe9302a2f3a31a23e8) until it's final.

Reviewer: Before this makes it to last call, the test cases should be inlined if at all possible.

Author: ok I'll provide that inlined before last call.

## Implementation
There is a various choice of existing implementations:
**Libraries:**
- Rust implementation (EY/Zexe): https://github.com/yelhousni/zexe/tree/youssef/BW6-761-Fq-ABLR-2ML-M
- C++ implementation (EY/libff): https://github.com/EYBlockchain/zk-swap-libff
- Golang implementation (Consensys/gurvy): https://github.com/ConsenSys/gurvy

**Stand-alone implementation:**
- Golang implementation with Intel assembly (Onur Kilic): https://github.com/kilic/bw6

**Precompiles:**
- OpenEthereum (EY/Parity): https://github.com/EYBlockchain/solidity-elliptic-curves

**Sripts:**

> - SageMath and Magma scripts: https://gitlab.inria.fr/zk-curves/bw6-761/

Reviewer: This section is no longer required.  Consider just including a reference implementation written in a language that is easy for a human to read and is self contained.  We try to avoid external links whenever possible since they tend to 404 with time, so an inline or `../assets/eip-3026/reference_implementation.whatever` would be preferred to an external implementation.

Author: This sounds good, but I would like to keep the implementations links for now until we agree upon a reference implementation. There are few implementations with different performances.
